Homework2/solution.cpp

```diff
@@ -0,0 +1,80 @@
+#include <iostream>
+#include <numeric>
+#include <thread>
+#include <vector>
+
+template<typename It, typename T>
+struct accumulator {
+    void operator()(It first, It last, T &result) {
+        result = std::accumulate(first, last, result);
+    }
+};
+
+void set_th_parameters(long length, unsigned int &th_quantity, long &block_size) {
+    auto hw_threads = std::thread::hardware_concurrency();
+    std::cout << "Threads used " << hw_threads << std::endl;
+    th_quantity = hw_threads != 0 ? hw_threads : 2;
+    block_size = length / th_quantity;
+}
+
+template<typename It, typename T>
+T th_accumulate(It first, It last, T init) {
+    auto length = std::distance(first, last);
+    if (!length) return init;
+
+    unsigned int th_quantity;
+    long block_size;
+    It block_start = first;
+
+    set_th_parameters(length, th_quantity, block_size);
+
+    std::vector<T> results(th_quantity);
+    std::vector<std::thread> threads(th_quantity - 1);
+
+    for (unsigned long i = 0; i < (th_quantity - 1); ++i) {
+        It block_end = block_start;
```

Comment:
to chyba może być ponad pętlą, bo później przy każdym obrocie linia 38 będzie robiła robotę. A tak to są dodatkowe operacje.